system-helper/flatpak-system-helper.c

```diff
@@ -1151,6 +1302,355 @@ handle_run_triggers (FlatpakSystemHelper   *object,
+    }
+
+  passwd = getpwnam ("flatpak");
+  if (passwd == NULL)
+    {
+      glnx_throw_errno_prefix (error, "User flatpak does not exist");
+      return FALSE;
+    }
+
+  return TRUE;
+}
+
+static void
+revokefs_fuse_backend_child_setup (gpointer user_data)
+{
+  struct passwd *passwd = getpwnam ("flatpak");
+  setgid (passwd->pw_gid);
+  setuid (passwd->pw_uid);
+}
+
+static void
+name_vanished_cb (GDBusConnection *connection, const gchar *name, gpointer user_data)
+{
+  OngoingPull *pull = (OngoingPull *) user_data;
+
+  ongoing_pull_free (pull);
+}
+
+static OngoingPull *
+ongoing_pull_new (FlatpakSystemHelper   *object,
+                  GDBusMethodInvocation *invocation,
+                  const gchar           *ref,
+                  const gchar           *commit,
+                  uid_t                  uid,
+                  const gchar           *src,
+                  GError               **error)
+{
+  GDBusConnection *connection = g_dbus_method_invocation_get_connection (invocation);
+  OngoingPull *pull;
+  g_autoptr (GSubprocessLauncher) launcher = NULL;
+  g_autoptr (GSubprocess) revokefs_backend = NULL;
+  g_autofree gchar *target_socket = NULL;
+  int sockets[2];
+
+  pull = g_slice_new0 (OngoingPull);
+  pull->object = object;
+  pull->invocation = invocation;
+  pull->ref = g_strdup (ref);
+  pull->commit = g_strdup (commit);
+  pull->src_dir = g_strdup (src);
+  pull->cancellable = g_cancellable_new ();
+  pull->uid = uid;
+  pull->preserve_pull = FALSE;
+
+  pull->watch_id = g_bus_watch_name_on_connection (connection,
+                                                   g_dbus_connection_get_unique_name (connection),
+                                                   G_BUS_NAME_WATCHER_FLAGS_NONE, NULL,
+                                                   name_vanished_cb,
+                                                   pull, NULL); /* TODO: Take a ref here */
+
+  if (socketpair (AF_UNIX, SOCK_SEQPACKET, 0, sockets) == -1)
+    {
+      glnx_throw_errno_prefix (error, "Failed to get a socketpair");
+      return NULL;
+    }
+
+  launcher = g_subprocess_launcher_new (G_SUBPROCESS_FLAGS_NONE);
+  g_subprocess_launcher_set_child_setup (launcher, revokefs_fuse_backend_child_setup, NULL, NULL);
+  g_subprocess_launcher_take_fd (launcher, sockets[0], 3);
+  fcntl (sockets[1], F_SETFD, FD_CLOEXEC);
+  target_socket = g_strdup ("--socket=3");
+  pull->revokefs_backend = g_subprocess_launcher_spawn (launcher,
+                                                        error,
+                                                        "revokefs-fuse",
+                                                        "--backend",
+                                                        target_socket, src, NULL);
+  if (pull->revokefs_backend == NULL)
+    return NULL;
+
+  /* Duplicate the client socket so that it can be actively shutdown during Deploy() */
+  pull->client_socket = fcntl(sockets[1], F_DUPFD_CLOEXEC, 0);
```

Comment:
Are you sure it's necessary to dup the file descriptor here? What's wrong with:
```suggestion
  pull->client_socket = sockets[1];
```

Otherwise, I think `sockets[1]` is leaked.

Reply:
See above in this function where `sockets` is declared. It doesn't need to be free-d. Makes sense? 
I don't think there is any other place where I should duplicate the client fd.

Reply:
`sockets` is not leaked but the file descriptor stored in `sockets[1]` is leaked. `sockets[0]` is not leaked because (per its [documentation](https://developer.gnome.org/gio/stable/GSubprocessLauncher.html#g-subprocess-launcher-take-fd)) `g_subprocess_launcher_take_fd()` takes ownership of the `fd` you pass it.